orchagent/vnetorch.cpp

```diff
@@ -1967,6 +2228,227 @@ void VNetRouteOrch::updateVnetTunnel(const BfdUpdate& update)
+                }
+            }
+            updateRoute = true;
+        }
+        if (updateRoute)
+        {
+            nhg_custom = nhg_custom_primary;
+        }
+    }
+    else if (nhg_custom_secondary.getSize() > 0)
+    {
+        if (nhg_custom_secondary != active_nhg )
+        {
+            if (!hasNextHopGroup(vnet, nhg_custom_secondary))
+            {
+                if (!createNextHopGroup(vnet, nhg_custom_secondary, vrf_obj, "custom"))
+                {
+                    SWSS_LOG_WARN("Failed to create primary based custom next hop group. Cannot proceed.");
+                    return;
+                }
+            }
+            updateRoute = true;
+        }
+        if (updateRoute)
+        {
+            nhg_custom = nhg_custom_secondary;
+        }
+    }
+    else
+    {
+        //both HHG's are inactive, need to remove the route.
+        updateRoute = true;
+    }
+
+    if (nhg_custom.getSize() == 0)
+    {
+        // nhg_custom is empty. we shall create a dummy empty NHG for book keeping.
+        SWSS_LOG_INFO(" Neither Primary or Secondary endpoints are up.");
+        if (!hasNextHopGroup(vnet, nhg_custom))
+        {
+            NextHopGroupInfo next_hop_group_entry;
+            next_hop_group_entry.next_hop_group_id = SAI_NULL_OBJECT_ID;
+            next_hop_group_entry.ref_count = 0;
+            syncd_nexthop_groups_[vnet][nhg_custom] = next_hop_group_entry;
+        }
+    }
+    auto active_nhg_size = active_nhg.getSize();
+    if (updateRoute)
+    {
+        for (auto vr_id : vr_set)
+        {
+            if (nhg_custom.getSize() == 0)
+            {
+                if (active_nhg_size > 0)
+                {
+                    // we need to remove the route
+                    del_route(vr_id, pfx);
+                }
+            }
+            else
+            {
+                bool route_status = true;
+                // note: nh_id can be SAI_NULL_OBJECT_ID when active_nhg is empty.
+                auto nh_id = syncd_nexthop_groups_[vnet][nhg_custom].next_hop_group_id;
+                if (active_nhg_size > 0)
+                {
+                    // we need to replace the nhg in the route
+                    route_status = update_route(vr_id, pfx, nh_id);
+                }
+                else
+                {
+                    // we need to readd the route.
+                    route_status = add_route(vr_id, pfx, nh_id);
+                }
+                if (!route_status)
+                {
+                    SWSS_LOG_ERROR("Route add/update failed for %s, vr_id '0x%" PRIx64, prefix.to_string().c_str(), vr_id);
+                    /* Clean up the newly created next hop group entry */
+                    if (nhg_custom.getSize() > 1)
+                    {
+                        removeNextHopGroup(vnet, nhg_custom, vrf_obj);
+                    }
+                    return;
+                }
+                vrf_obj->addRoute(prefix, nhg_custom);
```

Comment:
For route update (not add case), don't we need to also call removeRoute with active_nhg?

Reply:
we are already calling updateRoute.